tests/sound/sound.c

```diff
@@ -0,0 +1,278 @@
+
+    register int a0 asm("a0") = request_type;
+    register int a7 asm("a7") = 0xBABE;
+
+    asm volatile("scall" : "+r"(a0) : "r"(a7));
+}
+
+void shutdown_sound()
+{
+    int request_type = SHUTDOWN_SOUND;
+
+    register int a0 asm("a0") = request_type;
+    register int a7 asm("a7") = 0xBABE;
+
+    asm volatile("scall" : "+r"(a0) : "r"(a7));
+}
+
+/* use '=' to prevent ate by emulator */
+void usage(const char *prog)
+{
+    fprintf(
+        stderr,
+        "Usage: [path of rv32emu] %s [options]\n"
+        "Options:\n"
+        "  =m  [MUS format file]: convert MUS format to MIDI format and play "
+        "the music\n"
+        "  =s  [WAV format file]: play the sound effect\n"
+        "  =d  [music delay]    : the larger the more longer of playing the "
+        "music\n"
+        "  =mv [volume]         : set volume of music which in range [0 - 15], "
+        "default is 8\n"
+        "  =sv [volume]         : set volume of sfx which in range [0 - 15], "
+        "default is 15\n"
+        "  =upmv                : increasing music volume slowly to show "
+        "'set_music_volume' effect, \n"
+        "                         note: delay must be larger enough to see the "
+        "effect\n"
+        "  =srep                : repeat sfx sound during the play of music\n"
+        "  =h                   : show this usage\n",
+        prog);
+    exit(1);
+}
+
+/* FIXME: more robust parser of CLI arguments */
+static bool parse_args(int argc, char **argv)
+{
+    for (int i = 0; i < argc; i++) {
+        if (!strcmp(argv[i], "=m")) {
+            music_src = argv[i + 1];
+            has_music = true;
+        } else if (!strcmp(argv[i], "=s")) {
+            sfx_src = argv[i + 1];
+            has_sfx = true;
+        } else if (!strcmp(argv[i], "=d")) {
+            music_delay = (int) strtol(argv[i + 1], NULL, 10);
+        } else if (!strcmp(argv[i], "=upmv")) {
+            has_incr_music_volume = true;
+        } else if (!strcmp(argv[i], "=srep")) {
+            has_sfx_repeat = true;
+        } else if (!strcmp(argv[i], "=mv")) {
+            music_volume = (int) strtol(argv[i + 1], NULL, 10);
+        } else if (!strcmp(argv[i], "=sv")) {
+            sfx_volume = (int) strtol(argv[i + 1], NULL, 10);
+        } else if (!strcmp(argv[i], "=h")) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+/* FIXME: rough simulation of sleep */
+static void busy_loop(int n)
+{
+    int delay = n * 10000000;
+    int interval = delay / (15 - music_volume);
+
+    for (int i = 0; i < delay; i++) {
+        if (has_incr_music_volume && has_music) {
+            if (!(i % interval)) {
+                if (music_volume < 15) {
+                    music_volume++;
+                    set_music_volume();
+                }
+
+                if (has_sfx_repeat)
+                    play_sfx();
+            }
+        }
+    }
+}
+
+static void do_play_sound()
+{
+    if (has_music && has_sfx) {
```

Comment:
Don't do this! Employ bitwise tricks. See https://github.com/jserv/ir/blob/dev/tools/ir.c
e.g.,
```c
    if (dump & IR_DUMP_DUMP) {
        ir_dump(ctx, f);
    }
```